src/query_plan/query_planning_traversal.rs

```diff
@@ -86,3 +88,116 @@ struct BestQueryPlanInfo {
+            .parameters
+            .federated_query_graph
+            .root_kinds_to_nodes()?
+            .contains_key(&self.root_kind);
+
+        if is_root_path_tree {
+            // The root of the pathTree is one of the "fake" root of the subgraphs graph,
+            // which belongs to no subgraph but points to each ones.
+            // So we "unpack" the first level of the tree to find out our top level groups
+            // (and initialize our stack).
+            // Note that we can safely ignore the triggers of that first level
+            // as it will all be free transition, and we know we cannot have conditions.
+            for child in &path_tree.childs {
+                let edge = child.edge.expect("The root edge should not be None");
+                let (_source_node, target_node) = path_tree.graph.edge_endpoints(edge)?;
+                let target_node = path_tree.graph.node_weight(target_node)?;
+                let subgraph_name = &target_node.source;
+                // The edge tail type is one of the subgraph root type, so it has to be an ObjectType.
+                //     const rootType = edge.tail.type as ObjectType;
+                let root_type = match &target_node.type_ {
+                    QueryGraphNodeType::SchemaType(position) => match position {
+                        OutputTypeDefinitionPosition::Object(object) => object.clone().into(),
+                        _ => todo!(), // Return a FederationError?
+                    },
+                    QueryGraphNodeType::FederatedRootType(_) => todo!(), // Return a FederationError?
+                };
+                let group = dependency_graph.get_or_create_root_fetch_group(
+                    subgraph_name,
+                    self.root_kind,
+                    root_type,
+                );
+                compute_groups_for_tree(
+                    dependency_graph,
+                    &child.tree,
+                    // group,
+                    // GroupPath.empty(),
+                    // emptyDeferContext,
+                );
+            }
+        } else {
+            let query_graph_node = path_tree.graph.node_weight(path_tree.node)?;
+            let subgraph_name = &query_graph_node.source;
+            // The edge tail type is one of the subgraph root type, so it has to be an ObjectType.
+            let root_type = match &query_graph_node.type_ {
+                QueryGraphNodeType::SchemaType(position) => {
+                    position.clone().try_into().map_err(|err| {
+                        // TODO: add details to the error message?
+                        // format!("Should not have condition on non-selectable type {position}")
+                        err
+                    })?
+                }
+                QueryGraphNodeType::FederatedRootType(_) => todo!(), // Return a FederationError?
+            };
+            let group = dependency_graph.get_or_create_root_fetch_group(
```

Comment:
TypeScript function `computeNonRootFetchGroups` calls `getOrCreateRootFetchGroup`. Is it correct to mix "non-root" and "root" in this way?

Reply:
The gist is that the "root" vs. "non-root" in `computeRootFetchGroups()` and `computeNonRootFetchGroups()` is referring to whether the first node in the given `OpPathTree` is a federated root node or not. The "root" in `getOrCreateRootFetchGroup` means that the created fetch dependency graph node (i.e. "fetch group") is for a subgraph-entering transition and its fetch is against a root operation type of a subgraph (as opposed to something like `getOrCreateKeyFetchGroup()`, which is for a key resolution transition and its fetch is against `_entities`).